subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultImmutableAttributesContainer.java

```diff
@@ -29,67 +28,61 @@
 import org.jspecify.annotations.Nullable;
 
 import java.util.Comparator;
-import java.util.LinkedHashMap;
 import java.util.Map;
-import java.util.Objects;
 import java.util.TreeMap;
-import java.util.concurrent.ConcurrentHashMap;
 
 /**
- * Note: The fact that this type is both a <em>container</em> and a <em>value stored in a container</em> is strange.  This is
- * probably something that ought to be addressed, this shouldn't implement {@link AttributeValue}, but it does it in order to
- * build "linked lists" of immutable containers through concatenation to save memory.
+ * The default implementation of {@link ImmutableAttributes}.
+ * <p>
+ * Most methods are strict in that they will only return an entry if its
+ * key matches the requested key's name and type. However, {@link #getAttribute(Attribute)}
+ * attempts to coerce entries with a matching name to the requested type.
  */
-public final class DefaultImmutableAttributesContainer extends AbstractAttributeContainer implements ImmutableAttributes, AttributeValue<Object> {
-    private static final Comparator<Attribute<?>> ATTRIBUTE_NAME_COMPARATOR = Comparator.comparing(Attribute::getName);
-    // Coercion is an expensive process, so we cache the result of coercing to other attribute types.
-    // We can afford using a hashmap here because attributes are interned, and their lifetime doesn't
-    // exceed a build
-    private final Map<Attribute<?>, Object> coercionCache = new ConcurrentHashMap<>();
-
-    final Attribute<?> attribute;
-    final Isolatable<?> value;
-    private final ImmutableMap<Attribute<?>, DefaultImmutableAttributesContainer> hierarchy;
-    private final ImmutableMap<String, DefaultImmutableAttributesContainer> hierarchyByName;
+public final class DefaultImmutableAttributesContainer extends AbstractAttributeContainer implements ImmutableAttributes {
+
+    private final ImmutableAttributesEntry<?> head;
+    private final ImmutableMap<Attribute<?>, ImmutableAttributesEntry<?>> hierarchy;
+    private final ImmutableMap<String, ImmutableAttributesEntry<?>> hierarchyByName;
     private final int hashCode;
 
     // Optimize for the single entry case, makes findEntry faster
-    private final String singleEntryName;
-    private final DefaultImmutableAttributesContainer singleEntryValue;
-
-    DefaultImmutableAttributesContainer() {
-        this.attribute = null;
-        this.value = null;
-        this.hashCode = 0;
-        this.hierarchy = ImmutableMap.of();
-        this.hierarchyByName = ImmutableMap.of();
-        this.singleEntryName = null;
-        this.singleEntryValue = null;
-    }
+    private final @Nullable String singleEntryName;
+    private final @Nullable ImmutableAttributesEntry<?> singleEntry;
```

Comment:
❓ Is there still any real value in keeping these separate now?  Aren't these both just map lookups?  I doubt that optimizing for the case where a container stores only a single entry, and that happens to be the one searched for really helps.  Probably worth trying this change and measuring via the performance tests.

Reply:
I was also thinking about getting rid of this, but I really didn't want to mess with something with potential performance implications with this. 

I think it would be interesting to see if removing this actually makes any noticeable changes to performance. 

In practice, while map lookups are (pretty much) constant time, doing an `==` check is _vastly_ cheaper. Now whether that performance difference is worth the extra maintenance overhead of this extra code, thats different.